partiql-lang/src/main/kotlin/org/partiql/lang/planner/transforms/plan/PlanTyper.kt

```diff
@@ -155,6 +158,114 @@ internal object PlanTyper : PlanRewriter<PlanTyper.Context>() {
+                    val newType = excludeExprSteps(attr.type, expr.steps, lastStepAsOptional = false, ctx)
+                    resultAttrs.add(
+                        attr.copy(
+                            type = newType
+                        )
+                    )
+                }
+            } else {
+                resultAttrs.add(
+                    attr
+                )
+            }
+        }
+        return resultAttrs
+    }
+
+    private fun excludeExprSteps(type: StaticType, steps: List<ExcludeStep>, lastStepAsOptional: Boolean, ctx: Context): StaticType {
+        fun excludeExprStepsStruct(s: StructType, steps: List<ExcludeStep>, lastStepAsOptional: Boolean): StaticType {
+            val outputFields = mutableListOf<StructType.Field>()
+            val first = steps.first()
+            s.fields.forEach { field ->
+                when (first) {
+                    is ExcludeStep.TupleAttr -> {
+                        if (field.key == first.attr || (first.case == Case.INSENSITIVE && field.key.equals(first.attr, ignoreCase = true))) {
+                            if (steps.size == 1) {
+                                if (lastStepAsOptional) {
+                                    val newField = StructType.Field(field.key, field.value.asOptional())
+                                    outputFields.add(newField)
+                                }
+                            } else {
+                                outputFields.add(StructType.Field(field.key, excludeExprSteps(field.value, steps.drop(1), lastStepAsOptional, ctx)))
+                            }
+                        } else {
+                            outputFields.add(field)
+                        }
+                    }
+                    is ExcludeStep.TupleWildcard -> {
+                        if (steps.size == 1) {
+                            if (lastStepAsOptional) {
+                                val newField = StructType.Field(field.key, field.value.asOptional())
+                                outputFields.add(newField)
+                            }
+                        } else {
+                            outputFields.add(StructType.Field(field.key, excludeExprSteps(field.value, steps.drop(1), lastStepAsOptional, ctx)))
+                        }
+                    }
+                    else -> {
+                        handleInvalidExcludeExpr(ctx)
+                    }
+                }
+            }
+            return s.copy(fields = outputFields)
+        }
+
+        fun excludeExprStepsCollection(c: CollectionType, steps: List<ExcludeStep>, lastStepAsOptional: Boolean): StaticType {
+            var elementType = c.elementType
+            when (steps.first()) {
+                is ExcludeStep.CollectionIndex -> {
+                    if (steps.size > 1) {
+                        elementType = excludeExprSteps(elementType, steps.drop(1), lastStepAsOptional = true, ctx)
+                    }
```

Comment:
Sanity check here. 
So if we have `{'a' : [0, 1, 2]}` and exclude `a[0]`, the result will be `{'a' : [1,2]}`, hence optional, i.e., `unionOf(missing, int)`.

Reply:
Can you elaborate on this?

I would assume that the static type of `a` would be a `ListType(INT)`. IMO, even if we had:
```
SELECT * EXCLUDE t.a[0]
FROM << { 'a': [0] } >> AS t
```

I would assume the static type of `t.a` would still be `ListType(INT)` because we can't know how long the list actually is (especially if `t` is part of the catalog). If `t.a` started without any elements, I also think it would still be `ListType(INT)`. That being said @alancai98 , do we expect this to be a runtime error when we attempt to exclude an index that doesn't exist (essentially `OutOfBoundsException`)? How about when we try to exclude an attribute in a tuple that doesn't exist? This one we can sometimes know statically, but what happens at runtime if it's a union type?